src/Framework/IExtendedBuildEventArgs.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Build.Framework;
+
+/// <summary>
+/// Interface for Extended EventArgs to allow enriching particular events with extended data.
+/// Deriving from EventArgs will be deprecated soon and using Extended EventArgs is recommended for custom Event Args.
+/// </summary>
+public interface IExtendedBuildEventArgs
+{
+    /// <summary>
+    /// Unique string identifying type of extended data so receiver side knows how to interpret, deserialize and handle <see cref="ExtendedData"/>.
+    /// </summary>
+    string ExtendedType { get; set; }
+
+    /// <summary>
+    /// Metadata of <see cref="ExtendedData"/>.
+    /// Example usage:
+    ///   - data which needed in custom code to properly routing this message without interpreting/deserializing <see cref="ExtendedData"/>.
+    ///   - simple extended data can be transferred in form of dictionary key-value per one extended property.
+    /// </summary>
+    Dictionary<string, string?>? ExtendedMetadata { get; set; }
+
+    /// <summary>
+    /// Transparent data as string.
+    /// Custom code is responsible to serialize and deserialize this string to structured data - if needed.
+    /// Custom code can use any serialization they deem safe - e.g. json for textual data, base64 for binary data...
+    /// </summary>
+    string? ExtendedData { get; set; }
```

Comment:
What is the motivation for introducing this abstraction with Type, Data, and Metadata, instead of just `Dictionary<string, string?> ExtendedData`? It looks as though all of this is owned by the custom implementation (i.e. MSBuild doesn't interpret it) and simplifying it would make the hard-coded default `"undefined"` ExtendedType go away.

Reply:
It is based on my experience, opinionated, what shall generic purpose message/event have.

type => unique identifier of message content, for example cpp_error or my_structured_log_message
metadata => optional, describes data required to understand payload, for example `content_type=xml | json` or  `compression= none | gzip | deflate`
data => actual payload, in our case it is string as we expect mostly json, but people could base64 binaries if they want

Metadata can often be included into type like "cpp_error+json" but it is considered bad practice.

Similar approach of type+metadata+payload is used by many protocols including html, tcp/ip, amqp ...

I agree it is little bit overengineered, and people will usually not leverage it, but the price for including it was acceptable, IMO.

Reply:
Ok, since we can't extend interfaces later, overengineering makes sense I guess. If the parameterless constructors which set `ExtendedType` to the hardcoded `"undefined"` string are really for deserialization only, could they be made internal?